File: .claude/skills/sast-scan/rules/semgrep/go/tests/test_command_injection.go
package main

import (
	"context"
	"os/exec"
)

func vulnerable(userInput string) {
	// Positive: exec.Command with variable argument via sh -c
	// ruleid: go.security.command-injection-exec
	exec.Command("sh", "-c", userInput)

	// ruleid: go.security.command-injection-exec
	exec.Command("/bin/sh", "-c", userInput)

	// ruleid: go.security.command-injection-exec
	exec.Command("bash", "-c", userInput)

	// Positive: exec.CommandContext with variable
	// ruleid: go.security.command-injection-exec
	exec.CommandContext(context.Background(), "sh", "-c", userInput)

	// Positive: variable as command name
	// ruleid: go.security.command-injection-exec
	exec.Command(userInput, "-la")
}

func safe() {
	// Negative: literal command and arguments
	// ok: go.security.command-injection-exec
	exec.Command("ls", "-la")

	// ok: go.security.command-injection-exec
	exec.Command("ping", "-c", "3", "example.com")

	// ok: go.security.command-injection-exec
	exec.CommandContext(context.Background(), "echo", "hello")
}

File: .claude/skills/sast-scan/rules/semgrep/java/tests/TestCommandInjection.java
import java.io.IOException;

public class TestCommandInjection {
    public void vulnerable(String input) throws IOException {
        // Positive: Runtime.exec with string concatenation
        // ruleid: java.security.command-injection-runtime-exec
        Runtime.getRuntime().exec("sh -c " + input);

        // ruleid: java.security.command-injection-runtime-exec
        Runtime.getRuntime().exec("ping " + userInput);

        // ruleid: java.security.command-injection-runtime-exec
        Runtime.getRuntime().exec("curl " + url + " -o output.txt");

        // Positive: String.format in exec
        // ruleid: java.security.command-injection-runtime-exec
        Runtime.getRuntime().exec(String.format("sh -c %s", input));
    }

    public void safe(String[] args) throws IOException {
        // Negative: ProcessBuilder with array
        // ok: java.security.command-injection-runtime-exec
        ProcessBuilder pb = new ProcessBuilder("ls", "-la");
        pb.start();

        // ok: java.security.command-injection-runtime-exec
        ProcessBuilder pb2 = new ProcessBuilder(new String[]{"ping", "-c", "3", "example.com"});
        pb2.start();

        // ok: java.security.command-injection-runtime-exec
        Runtime.getRuntime().exec(new String[]{"ls", "-la"});
    }
}
